modules/images/dominant-color/load.php

```diff
@@ -0,0 +1,425 @@
+ * Experimental: Yes
+ *
+ * @package performance-lab
+ * @since 1.0.0
+ */
+
+/**
+ * Add Dominant Color preload support to WordPress.
+ *
+ * TODO: Add tests
+ */
+class wp_Dominant_Color {
+
+
+	public function __construct() {
+
+		add_filter( 'wp_print_scripts', [ $this, 'add_styles' ] );
+		add_filter( 'wp_footer', [ $this, 'add_scripts' ] );
+
+		add_filter( 'wp_generate_attachment_metadata', [ $this, 'dominant_color_metadata' ], 10, 2 );
+		add_filter( 'wp_generate_attachment_metadata', [ $this, 'has_transparency_metadata' ], 10, 2 );
+
+		// do we have the new filter or are duplicating core the functions?
+		if ( has_filter( 'wp_img_tag_add_adjust' ) ) {
+			add_filter( 'wp_img_tag_add_adjust', [ $this, 'tag_add_adjust' ], 20, 3 );
+		} else {
+			add_filter( 'the_content', [ $this, 'filter_content_tags' ], 20 );
+			add_filter( 'wp_dominant_color_img_tag_add_adjust', [ $this, 'tag_add_adjust' ], 10, 3 );
+		}
+
+		add_filter( 'attachment_fields_to_edit', [ $this, 'add_dominant_color_media_setting' ], 10, 2 );
+		add_filter( 'attachment_fields_to_save', [ $this, 'save_dominant_color_media_setting' ], 10, 2 );
+	}
+
+	/**
+	 * Add the dominant color metadata to the attachment.
+	 *
+	 * @param array $metadata
+	 * @param int $attachment_id
+	 *
+	 * @return array $metadata
+	 */
+	public function dominant_color_metadata( $metadata, $attachment_id ) {
+
+		$dominant_color = $this->get_dominant_color( $attachment_id );
+
+		if ( ! empty( $dominant_color ) ) {
+			$metadata['dominant_color'] = $dominant_color;
+		}
+
+		return $metadata;
+	}
+
+	/**
+	 * Add the dominant color metadata to the attachment.
+	 *
+	 * @param array $metadata
+	 * @param int $attachment_id
+	 *
+	 * @return array $metadata
+	 */
+	public function has_transparency_metadata( $metadata, $attachment_id ) {
+
+		$has_transparency = $this->get_has_transparency( $attachment_id );
+
+		if ( ! empty( $has_transparency ) ) {
+			$metadata['has_transparency'] = $has_transparency;
+		}
+
+		return $metadata;
+	}
+
+	/**
+	 * @param $filtered_image
+	 * @param $context
+	 * @param $attachment_id
+	 *
+	 * @return string image tag
+	 */
+	public function tag_add_adjust( $filtered_image, $context, $attachment_id ) {
+
+		$image_meta = wp_get_attachment_metadata( $attachment_id );
+
+		if ( isset( $image_meta['dominant_color'] ) ) {
+			$data  = sprintf( 'data-dominantColor="%s"', $image_meta['dominant_color'] );
+			$style = '';
+			if ( strpos( $filtered_image, 'loading="lazy"' ) !== false && ( isset( $image_meta['show_dominant_color'] ) && 1 === (int) $image_meta['show_dominant_color'] ) ) {
+				$style = ' style="--dominant-color: #' . $image_meta['dominant_color'] . '; " ';
+			}
+
+			$extra_class = '';
+
+			if ( ! isset( $image_meta['has_transparency'] ) ) {
+				$data        .= ' data-hasTransparency="false"';
+			} else {
+				$data .= ' data-hasTransparency="true"';
```

Comment:
Here too, don't use the camel-case in the HTML attributes.
Here's the why
https://usefulangle.com/post/172/javascript-access-data-attributes

`data-post-id=`

is accessed in JS as
`.dataset.postId`